x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStopDatafeedAction.java

```diff
@@ -363,16 +379,33 @@ private void sendResponseOrFailure(String datafeedId, ActionListener<StopDatafee
         listener.onFailure(e);
     }
 
-    // Wait for datafeed to be marked as stopped in cluster state, which means the datafeed persistent task has been removed
-    // This api returns when task has been cancelled, but that doesn't mean the persistent task has been removed from cluster state,
-    // so wait for that to happen here.
-    void waitForDatafeedStopped(List<String> datafeedPersistentTaskIds, StopDatafeedAction.Request request,
+    /**
+     * Wait for datafeed to be marked as stopped in cluster state, which means the datafeed persistent task has been removed.
+     * This api returns when task has been cancelled, but that doesn't mean the persistent task has been removed from cluster state,
+     * so wait for that to happen here.
+     *
+     * Since the stop datafeed action consists of a chain of async callbacks, it's possible that datafeeds have moved nodes since we
+     * decided what to do with them at the beginning of the chain.  We cannot simply wait for these, as the request to stop them will
+     * have been sent to the wrong node and ignored there, so we'll just spin until the timeout expires.
```

Comment:
Excellent catch here. Yeah, this has always been a weird edge case. Glad its found now :)